### PR TITLE
Add generate_series() udtf (and introduce 'lazy' `MemoryExec`)

### PR DESCRIPTION
**Pull request by 2010YOUY01**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10069

## Rationale for this change

To make it easier to test memory-limited queries, i was looking for a datasource which can generate long streams, and itself doesn't consume lots of memory.
There are several existing approaches are close, but can't fully satisfy this requirement:
- `MemoryExec` have to buffer all output data up front
- `select * from unnest(generate_series(1,10000))` also have to buffer all output during construction
- It's possible to do such tests on large files, however it would be hard to setup and takes longer to run

So in this PR, a new `StreamingMemoryExec` is introduced, it can be used to generate batches lazily by defining a 'iterator' on `RecordBatch` in its interface. And a new user-defined table function is implemented with this new execution plan.

This function's behavior is kept consistent with [DuckDB's generate_series() table function](https://duckdb.org/docs/sql/functions/list#generate_series), and only 2 argument variant is implemented for now

I also think this UDTF can be useful to tests/micro-benchmarks in general

```
> select * from generate_series(1,3);
+-------+
| value |
+-------+
| 1     |
| 2     |
| 3     |
+-------+
```

## What changes are included in this PR?

1. Move `TableFunctionImpl` from `datafusion` crate -> `datafusion-catalog` crate, so new UDTFs defined in `datafusion-functions-table` crate won't have circular dependency with `datafusion` crate (`TableFunctionImpl` depends on `TableProvider` in `datafusion-catalog` crate, so I think it's the best place to move it to)
2. Introduced `StreamingMemoryExec`
3. Implemented a new UDTF `generate_series()`

## Are these changes tested?

Unit tests for `StreamingMemoryExec`, and `sqllogictest`s for `generate_series()` UDTF

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/physical-plan/src/memory.rs [hunk 1]
```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
...
+
+impl StreamingMemoryExec {
+    /// Create a new streaming memory execution plan
+    pub fn try_new(
+        schema: SchemaRef,
+        generators: Vec<Box<dyn StreamingBatchGenerator>>,
+    ) -> Result<Self> {
+        let cache = PlanProperties::new(
+            EquivalenceProperties::new(Arc::clone(&schema)),
+            Partitioning::RoundRobinBatch(generators.len()),
```
Comment (2010YOUY01) [inline datafusion/physical-plan/src/memory.rs:401]: Is this field mean: Let's say a exec only have 1 generator, the optimizer will try to insert a `RepartitionExec` to the output of `StreamingMemoryExec`, and use round robin to repartition output batches to `target_partitions` number?
I found it works on some aggregate queries, but not for a sort query
```
> explain select * from generate_series(1, 10000) as t1(v1) order by v1;
+---------------+----------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                           |
+---------------+----------------------------------------------------------------------------------------------------------------+
| logical_plan  | Sort: t1.v1 ASC NULLS LAST                                                                                     |
|               |   SubqueryAlias: t1                                                                                            |
|               |     Projection: tmp_table.value AS v1                                                                          |
|               |       TableScan: tmp_table projection=[value]                                                                  |
| physical_plan | SortExec: expr=[v1@0 ASC NULLS LAST], preserve_partitioning=[false]                                            |
|               |   ProjectionExec: expr=[value@0 as v1]                                                                         |
|               |     StreamingMemoryExec: partitions=1, batch_generators=[generate_series: start=1, end=10000, batch_size=8192] |
|               |                                                                                                                |
+---------------+----------------------------------------------------------------------------------------------------------------+
```
I want to know is it correct here, and this sort query in theory can be made parallelized by changing somewhere in optimization phase? 🤔

File: datafusion/physical-plan/src/memory.rs [hunk 2]
```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
...
+
+    fn statistics(&self) -> Result<Statistics> {
+        Ok(Statistics::new_unknown(&self.schema))
+    }
+}
+
+/// Stream that generates record batches on demand
+pub struct StreamingMemoryStream {
+    schema: SchemaRef,
+    generator: Box<dyn StreamingBatchGenerator>,
```
Comment (jayzhan211) [inline datafusion/physical-plan/src/memory.rs:499]: Should we use `Arc` given it makes sense to have generator generate across threads

File: datafusion/physical-plan/src/memory.rs [hunk 3]
```diff
@@ -365,8 +365,165 @@ impl RecordBatchStream for MemoryStream {
...
+
+/// Execution plan for streaming in-memory batches of data
+///
+/// This plan generates output batches lazily, it doesn't have to buffer all batches
+/// in memory up front (compared to `MemoryExec`), thus consuming constant memory.
+pub struct StreamingMemoryExec {
+    /// Schema representing the data
+    schema: SchemaRef,
+    /// Functions to generate batches for each partition
+    batch_generators: Vec<Box<dyn StreamingBatchGenerator>>,
```
Comment (jayzhan211) [inline datafusion/physical-plan/src/memory.rs:388]:
```suggestion
    batch_generators: Vec<Arc<dyn StreamingBatchGenerator>>,
```

File: datafusion/catalog/src/table.rs [hunk 4]
```diff
@@ -297,3 +299,40 @@ pub trait TableProviderFactory: Debug + Sync + Send {
         cmd: &CreateExternalTable,
     ) -> Result<Arc<dyn TableProvider>>;
 }
+
+/// A trait for table function implementations
+pub trait TableFunctionImpl: Debug + Sync + Send {
```
Comment (jayzhan211) [inline datafusion/catalog/src/table.rs:304]:
```suggestion
/// Returns this function's name
    fn name(&self) -> &str;
```

This is much more consistent with other UDFs

File: datafusion/sqllogictest/test_files/table_functions.slt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+EXPLAIN SELECT * FROM generate_series(1, 5)
+----
+logical_plan TableScan: tmp_table projection=[value]
+physical_plan StreamingMemoryExec: partitions=1, batch_generators=[generate_series: start=1, end=5, batch_size=8192]
+
+query error DataFusion error: This feature is not implemented: End value must be greater than or equal to start value
+SELECT * FROM generate_series(5, 1)
+
+statement error DataFusion error: Error during planning: First argument must be an integer literal
+SELECT * FROM generate_series(NULL, 5)
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/table_functions.slt:106]: We might not return error for this query
```
D SELECT * FROM generate_series(null, 5);
┌─────────────────┐
│ generate_series │
│      int64      │
├─────────────────┤
│     0 rows      │
└─────────────────┘
```

File: datafusion/sqllogictest/test_files/table_functions.slt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+statement error DataFusion error: Error during planning: Second argument must be an integer literal
+SELECT * FROM generate_series(1, NULL)
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/table_functions.slt:109]:
```
D SELECT * FROM generate_series(1, null);
┌─────────────────┐
│ generate_series │
│      int64      │
├─────────────────┤
│     0 rows      │
└─────────────────┘
```

File: datafusion/sqllogictest/test_files/table_functions.slt [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+statement error DataFusion error: Error during planning: generate_series expects 2 arguments
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/table_functions.slt:111]:
```
D SELECT * FROM generate_series(1, 5, null);
┌─────────────────┐
│ generate_series │
│      int64      │
├─────────────────┤
│     0 rows      │
└─────────────────┘
```

File: datafusion/sqllogictest/test_files/table_functions.slt (hunk 5 above)
Comment (2010YOUY01) [reply datafusion/sqllogictest/test_files/table_functions.slt:106]: Agreed, it's a good idea to align the behavior with DuckDB

File: datafusion/catalog/src/table.rs (hunk 4 above)
Comment (2010YOUY01) [reply datafusion/catalog/src/table.rs:304]: This is the part of the refactor, changing this requires several fixes. We can leave it to a separate PR to make this PR smaller

File: datafusion/sqllogictest/test_files/table_functions.slt (hunk 7 above)
Comment (2010YOUY01) [reply datafusion/sqllogictest/test_files/table_functions.slt]: I'll open an issue for 3-arg support later

**Comment by 2010YOUY01** [comment]
Thank you for the comments @jayzhan211 , I have updated.

Now I think the concurrent generator might not be very straightforward, I'll first leave some rationale here.
If we can come to agreement I'll add more doc to it (else revert back to `Box` version for simplicity)

----
### Rationale for `StreamingMemoryStream` design
Suppose we want to run `select ... from generate_series(1,100)` in two partitions
And the underlying batch generator is wrapped with Mutex
```
pub struct StreamingMemoryStream {
    ...
    generator: Arc<Mutex<dyn StreamingBatchGenerator>>,
}
```
It's possible to implement the UDTF in 3 ways:
1.
```
[generator_1_50]   --- [StreamingMemoryStream  stream1] --> xxStream1
[generator_50_100] --- [StreamingMemoryStream  stream2] --> xxStream1
```
2.
```
[generator_1_100] --- [StreamingMemoryStream  stream1] --> Repartition --> xxStream1
                                                                       |-> xxStream2                                                                                
```
3.
```
[generator_1_100] --- [StreamingMemoryStream  stream1] --> xxStream1
                  |-- [StreamingMemoryStream  stream2] --> xxStream2                                                                                                                                      
```
1 and 2 is the common pattern for datafusion scanning operators to do plan-time parallelism, `Mutex` is not used in this case
3 make the `StreamingBatchGenerator` being able to concurrently accessed by multiple streams. It's trying to make the interface more general-purpose for future usage, and won't be used by `generate_series()`

**Comment by jayzhan211** [comment]
> Thank you for the comments @jayzhan211 , I have updated.
> 
> Now I think the concurrent generator might not be very straightforward, I'll first leave some rationale here. If we can come to agreement I'll add more doc to it (else revert back to `Box` version for simplicity)
> 
> ### Rationale for `StreamingMemoryStream` design
> Suppose we want to run `select ... from generate_series(1,100)` in two partitions And the underlying batch generator is wrapped with Mutex
> 
> ```
> pub struct StreamingMemoryStream {
>     ...
>     generator: Arc<Mutex<dyn StreamingBatchGenerator>>,
> }
> ```
> 
> It's possible to implement the UDTF in 3 ways: 1.
> 
> ```
> [generator_1_50]   --- [StreamingMemoryStream  stream1] --> xxStream1
> [generator_50_100] --- [StreamingMemoryStream  stream2] --> xxStream1
> ```
> 
> 
> ```
> [generator_1_100] --- [StreamingMemoryStream  stream1] --> Repartition --> xxStream1
>                                                                        |-> xxStream2                                                                                
> ```
> 
> 
> ```
> [generator_1_100] --- [StreamingMemoryStream  stream1] --> xxStream1
>                   |-- [StreamingMemoryStream  stream2] --> xxStream2                                                                                                                                      
> ```
> 
> 1 and 2 is the common pattern for datafusion scanning operators to do plan-time parallelism, `generator` won't be accessed by multiple threads thus `Mutex` is redundant 3 make the `StreamingBatchGenerator` being able to concurrently accessed by multiple streams.
> 
> The `Mutex` is added to make it possible for case 3 (so the interface can be more general-purpose for future use cases)

Given the quick look, I'm not sure whether we need 3rd case. It seems the 1st case runs execution parallelly too. I would need to think about the advantage of the 3rd case over 1st case.

For 3rd case, if we need it, we might use `Arc<RwLock<dyn T>>` if `generate_next_batch` takes `&self`. It might be more efficiently.

File: datafusion/catalog/src/table.rs (hunk 4 above)
Comment (berkaysynnada) [reply datafusion/catalog/src/table.rs:304]: I also recommend to take `name()` inside `TableFunctionImpl`

File: datafusion/core/src/execution/session_state_defaults.rs [hunk 8]
```diff
@@ -119,6 +120,11 @@ impl SessionStateDefaults {
         functions_window::all_default_window_functions()
     }
 
+    /// returns the list of default [`TableFunction`]s
+    pub fn default_table_functions() -> Vec<Arc<TableFunction>> {
+        functions_table::all_default_table_functions()
+    }
+
```
Comment (berkaysynnada) [inline datafusion/core/src/execution/session_state_defaults.rs:127]: Can we get the defaults from a singleton ? Like other default getters, with get_or_init or smth similar

File: datafusion/functions-table/src/generate_series.rs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    start: Option<i64>,
+    // None if input is Null
+    end: Option<i64>,
+}
+
+/// Table state that generates a series of integers from `start`(inclusive) to `end`(inclusive)
+#[derive(Debug, Clone)]
+struct GenerateSeriesState {
+    schema: SchemaRef,
+    _start: i64,
```
Comment (berkaysynnada) [inline datafusion/functions-table/src/generate_series.rs:47]: Is there any reason to keep this?

File: datafusion/functions-table/src/generate_series.rs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+#[derive(Debug)]
+pub struct GenerateSeriesFunc {}
+
+impl TableFunctionImpl for GenerateSeriesFunc {
+    // Check input `exprs` type and number. Input validity check (e.g. start <= end)
+    // will be performed in `TableProvider::scan`
+    fn call(&self, exprs: &[Expr]) -> Result<Arc<dyn TableProvider>> {
+        // TODO: support 3 arguments following DuckDB:
```
Comment (berkaysynnada) [inline datafusion/functions-table/src/generate_series.rs:150]: There is also a one length argument usage, perhaps can be added as todo

File: datafusion/physical-plan/src/memory.rs [hunk 11]
```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+pub trait StreamingBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
```
Comment (berkaysynnada) [inline datafusion/physical-plan/src/memory.rs:369]: I am not very satisfied with these `Streaming... `naming. Can we describe the behavior better with a different naming for these? Perhaps we could use 'lazy' term

**Review by berkaysynnada — APPROVED** [review]
Thank you @2010YOUY01. I reviewed the changes and LGTM. I have a few minor comments and one question: I noticed another approach of `generate_series()`, which can be used like this: 
`SELECT generate_series(1, 5)` 
I assume it is not a udtf in that context. Does this implementation leave room for that usage?

File: datafusion/core/src/execution/session_state_defaults.rs (hunk 8 above)
Comment (2010YOUY01) [reply datafusion/core/src/execution/session_state_defaults.rs:127]: This is a good idea, updated

File: datafusion/functions-table/src/generate_series.rs (hunk 9 above)
Comment (2010YOUY01) [reply datafusion/functions-table/src/generate_series.rs]: This field is for display, i added a comment to explain

**Comment by 2010YOUY01** [comment]
> Thank you @2010YOUY01. I reviewed the changes and LGTM. I have a few minor comments and one question: I noticed another approach of `generate_series()`, which can be used like this: `SELECT generate_series(1, 5)` I assume it is not a udtf in that context. Does this implementation leave room for that usage?

Thank you for the review @berkaysynnada 

The answer to the question is yes, they are different function with the same name.
Using same name is probably not the best idea, but we do this to keep the behavior the same as DuckDB.
I added a sql test to make sure they can work together (like `select generate_series() from generate_series()`)

----
@jayzhan211 I thought about the sharing generator issue again, I think making a interface more flexible is important, so I kept the lock, and updated `RwLock` as you suggested

File: Cargo.toml [hunk 12]
```diff
@@ -30,6 +30,7 @@ members = [
     "datafusion/functions",
     "datafusion/functions-aggregate",
     "datafusion/functions-aggregate-common",
+    "datafusion/functions-table",
```
Comment (alamb) [inline Cargo.toml:33]: 100%

File: datafusion/catalog/src/table.rs (hunk 4 above)
Comment (alamb) [reply datafusion/catalog/src/table.rs:304]: I think doing it as a follow on PR makes sense to me -- perhaps we can file a ticket

Another thing that would be nice for a follow on PR is to move this trait into its own module (`catalog/src/table_function.rs` perhaps)

File: datafusion/functions-table/src/generate_series.rs [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    start: Option<i64>,
+    // None if input is Null
+    end: Option<i64>,
+}
+
+/// Table state that generates a series of integers from `start`(inclusive) to `end`(inclusive)
+#[derive(Debug, Clone)]
+struct GenerateSeriesState {
+    schema: SchemaRef,
+    _start: i64, // Kept for display
```
Comment (alamb) [inline datafusion/functions-table/src/generate_series.rs:47]: does it need to be prefixed with `_` if it is used for display 🤔

File: datafusion/functions-table/src/generate_series.rs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+#[derive(Debug)]
+pub struct GenerateSeriesFunc {}
+
+impl TableFunctionImpl for GenerateSeriesFunc {
+    // Check input `exprs` type and number. Input validity check (e.g. start <= end)
+    // will be performed in `TableProvider::scan`
+    fn call(&self, exprs: &[Expr]) -> Result<Arc<dyn TableProvider>> {
+        // TODO: support 1 or 3 arguments following DuckDB:
```
Comment (alamb) [inline datafusion/functions-table/src/generate_series.rs:150]: can we possibly file a ticket to track this?

File: datafusion/physical-plan/src/memory.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-plan/src/memory.rs:395]: Maybe you can look at https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.repartitioned or similar methods 🤔

File: datafusion/physical-plan/src/memory.rs [hunk 15]
```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
...
+/// Execution plan for lazy in-memory batches of data
+///
+/// This plan generates output batches lazily, it doesn't have to buffer all batches
+/// in memory up front (compared to `MemoryExec`), thus consuming constant memory.
+pub struct LazyMemoryExec {
+    /// Schema representing the data
+    schema: SchemaRef,
+    /// Functions to generate batches for each partition
+    batch_generators: Vec<Arc<RwLock<dyn LazyBatchGenerator>>>,
+    /// Total number of rows to generate for statistics
```
Comment (alamb) [inline datafusion/physical-plan/src/memory.rs:383]: this comment doesn't seem quite right

**Review by alamb — APPROVED** [review]
Thank ou @2010YOUY01 @jayzhan211 and @berkaysynnada 

I left some structural comments / thoughts about how we might be able to make LazyMemoryExec more general (`StreamExec` or `StreamWrapperExec`) but in general I think this is a really nice first step and we could do additional refactorings as follow on PRs.

File: datafusion/physical-plan/src/memory.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/physical-plan/src/memory.rs:395]: I'm not sure about this too

**Review by jayzhan211 — APPROVED** [review]

File: datafusion/physical-plan/src/memory.rs [hunk 16]
```diff
@@ -365,8 +366,165 @@ impl RecordBatchStream for MemoryStream {
     }
 }
 
+pub trait LazyBatchGenerator: Send + Sync + fmt::Debug + fmt::Display {
```
[the remark this answers is not in the document]
Comment (2010YOUY01) [reply datafusion/physical-plan/src/memory.rs:369]: This is a really good idea, filed https://github.com/apache/datafusion/issues/13614

File: datafusion/functions-table/src/generate_series.rs (hunk 14 above)
Comment (2010YOUY01) [reply datafusion/functions-table/src/generate_series.rs:150]: Tracked in https://github.com/apache/datafusion/issues/13615

File: datafusion/catalog/src/table.rs (hunk 4 above)
Comment (2010YOUY01) [reply datafusion/catalog/src/table.rs:304]: Filed https://github.com/apache/datafusion/issues/13613

**Comment by 2010YOUY01** [comment]
Thank you @alamb , I have updated minor renaming/doc fixes in the latest commit, and opened issues for all other bigger follow on tasks

**Comment by alamb** [comment]
Looks good -- let's get this merged 🚀

**Pull request merged**